Run information for co-added runs in MuonAnalysis

tom-perkins opened the pull request:
Fixes the run information displayed in the MuonAnalysis interface when several runs are co-added.

Corrects each point raised in the issue:
- "Runs" deals with non-consecutive ranges and gaps
- "Sample temperature" and "sample magnetic field" are shown as a range if they are different for each run
- "Average temperature" is derived from all the runs
- "Start" and "End" are the earliest start and latest end

Previously these fields showed data which applied only to the first run.

The new functionality is in `MuonAnalysisHelper` rather than the main MuonAnalysis class, so it was possible to add unit tests. Each new piece of functionality should have a test.

**To test:**
_Code review:_ 
Ensure the tests cover the new functionality

_Functional tests:_
Files `MUSR15189.nxs` to `MUSR15193.nxs` can be found in the DocTest folder.
- Start MuonAnalysis interface
- Load a selection of these files by typing a range of run numbers into the box
- Verify "runs" is shown correctly
- Check the start and end times, and range of magnetic fields
- Try a sequence of files with a gap e.g. 15189-91, 15193
- To test the sample temperature/average temperature, different files are needed. Try MUSR24580-1 in the training data (can be found at `\\olympic\babylon5\Scratch\TomPerkins\Data`) to see a range of sample temperatures, and check the average temperature is derived from both files.
- Test the above when you have the default instrument in Mantid (Help/First time setup) set to a different instrument than the one selected in the interface

Fixes #13738 

The [release notes](https://github.com/mantidproject/mantid/blob/13738_Run_info_MuonAnalysis/docs/source/release/v3.7.0/muon.rst) have been updated.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

KarlPalmen commented on MantidQt/CustomInterfaces/test/MuonAnalysisHelperTest.h:
```diff
@@ -124,6 +132,47 @@ class MuonAnalysisHelperTest : public CxxTest::TestSuite {
+  void test_findStartAndEndTimes_singleWS() {
+    Workspace_sptr ws = createWs("MUSR", 15189);
+    DateAndTime start{"2015-12-23T15:32:40Z"};
+    DateAndTime end{"2015-12-24T09:00:00Z"};
```
I'd call them expectedStart and expectedEnd.

KarlPalmen commented:
When attempting to load "15189-91, 15193" I get message "Unable to find file: Search object: 15189", but "15189-15191" works OK and also "15193" and "15189-19193" work OK and also selecting multiple files in browse works OK.

tom-perkins commented:
I see the same error when the instrument set on the "First time setup" screen is not MUSR. If I then set it to MUSR via Help/First time setup, it works. 

I think this is a bug because it should use the specific instrument set in the interface, not the general one set in Mantid. (I don't understand why it will load a continuous range of files but not one with gaps in this case!) 

I'll change the 'load files' box to search for files using the correct instrument and push a fix.

KarlPalmen commented:
Yes. This needs reporting as a bug. It also works for two non-consecutive individual files. It is only the mixed format that fails.

KarlPalmen commented:
I'll soon be taking annual leave from tomorrow till after Easter, so will leave it for someone else to test.

tom-perkins commented:
Added a fix for the problem, which was that `MWRunFiles` can cope with ranges and lists but not a list with ranges in it. Modified the tests in the top comment to add the special case.

KarlPalmen commented:
Test now works with mixed list and ranges and still works with ranges and lists.

KarlPalmen commented:
:shipit:

FedeMPouzols commented:
This looks good and ready to ship, but @KarlPalmen can you tick the relevant checkboxes in the PR description? I think in this PR many of them are applicable.

KarlPalmen commented:
I've ticked some check boxes. 

I now realise I did not see the addition of a unit test for the mixed list and range case.
